src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/ModelExpressionProvider.cs

```diff
@@ -0,0 +1,76 @@
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Linq.Expressions;
+using Microsoft.AspNetCore.Mvc.ModelBinding;
+using Microsoft.AspNetCore.Mvc.ViewFeatures;
+using Microsoft.AspNetCore.Mvc.ViewFeatures.Internal;
+
+namespace Microsoft.AspNetCore.Mvc.Rendering
+{
+    public class ModelExpressionProvider : IModelExpressionProvider
+    {
+        private readonly ExpressionTextCache _expressionTextCache;
+        private readonly IModelMetadataProvider _modelMetadataProvider;
+
+        public ModelExpressionProvider(
+            IModelMetadataProvider metadataProvider,
+            ExpressionTextCache expressionTextCache
+            )
+        {
+            if (metadataProvider == null)
+            {
+                throw new ArgumentNullException(nameof(metadataProvider));
+            }
+
+            if (expressionTextCache == null)
+            {
+                throw new ArgumentNullException(nameof(expressionTextCache));
+            }
+
+            _modelMetadataProvider = metadataProvider;
+            _expressionTextCache = expressionTextCache;
+        }
+
+        /// <summary>
+        /// Returns a <see cref="ModelExpression"/> instance describing the given <paramref name="expression"/>.
+        /// </summary>
+        /// <typeparam name="TValue">The type of the <paramref name="expression"/> result.</typeparam>
+        /// <typeparam name="TModel">The type of the <paramref name="viewData"/> result.</typeparam>
+        /// <param name="expression">An expression to be evaluated against the current model.</param>
+        /// <param name="viewData">
+        /// The <see cref="ViewDataDictionary"/> that may contain the <paramref name="expression"/> value.
+        /// </param>
+        /// <returns>A new <see cref="ModelExpression"/> instance describing the given <paramref name="expression"/>.
+        /// </returns>
+        /// <remarks>
+        /// Compiler normally infers <typeparamref name="TValue"/> from the given <paramref name="expression"/>.
+        /// </remarks>
+        public ModelExpression CreateModelExpression<TModel, TValue>(
+            ViewDataDictionary<TModel> viewData,
+            Expression<Func<TModel, TValue>> expression)
+        {
+            if (expression == null)
+            {
+                throw new ArgumentNullException(nameof(expression));
+            }
+
+            if (viewData == null)
+            {
+                throw new ArgumentNullException(nameof(viewData));
+            }
+
+            var name = ExpressionHelper.GetExpressionText(expression, _expressionTextCache);
+            var modelExplorer = ExpressionMetadataProvider.FromLambdaExpression(expression, viewData, _modelMetadataProvider);
+            if (modelExplorer == null)
+            {
+                // This message has to be created in the resources file of the new namespace.
```

Comment:
What you need to do is move the resource from `src/Microsoft.AspNetCore.Mvc.Razor/Resources.resx` to `src/Microsoft.AspNetCore.Mvc.ViewFeatures/Resources.resx`. Then the `FormatRazorPage_NullModelMetadata()` should be in a namespace this class is already `using`.